File: kotlin/src/main/kotlin/ListNode.kt
package o.mg

class ListNode(var `val`: Int) {
    var next: ListNode? = null

    private fun toList(): List<Int> {
        var node: ListNode? = this
        val nodes = mutableListOf<Int>()
        while (node != null) {
            nodes.add(node.`val`)
            node = node.next
        }
        return nodes.toList()
    }

    override fun toString(): String {
        return this.toList().joinToString(separator = ",", prefix = "<", postfix = ">")
    }

    override fun equals(other: Any?): Boolean {
        if (other !is ListNode?) {
            return false
        }
        if (other == null) return true
        if (this === other) return true
        if (this.`val` == other.`val`) {
            return this.next == other.next
        }
        return false
    }

    constructor(list: List<Int>) : this(list.first()) {
        val v: ListNode? = list.reversed().fold(null) { acc: ListNode?, i ->
            val c = ListNode(i)
            c.next = acc
            c
        }
        this.next = v!!.next
    }
}

File: python/src/list_node.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __eq__(self, value) -> bool:
        if self is None and value is None:
            return True
        if self is None or value is None:
            return False
        if self is value:
            return True
        if self.val != value.val:
            return False
        return self.next == value.next

    def __str__(self) -> str:
        if self is None:
            return "None"
        return f"<{self.val}, {self.next}>"
